File: protocol/src/main/resources/lua/ProtocolTemplate.lua
{}
local {} = {}

function {}:new()
    local obj = {
        {}
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function {}:protocolId()
    return {}
end

function {}:protocolName()
    return {}
end

function {}:__tostring()
    local jsonTemplate = "{}"
    local result = string.format(jsonTemplate, {})
    return result
end

function {}:write(buffer, packet)
    if packet == nil then
        buffer:writeInt(0)
        return
    end
    {}
end

function {}:read(buffer)
    local length = buffer:readInt()
    if length == 0 then
        return nil
    end
    local beforeReadIndex = buffer:getReadOffset()
    local packet = {}:new()
    {}
    if length > 0 then
        buffer:setReadOffset(beforeReadIndex + length)
    end
    return packet
end

return {}

File: protocol/src/main/resources/typescript/ProtocolTemplate.ts
{}
{}
class {} {

    {}

    static PROTOCOL_ID: number = {};

    protocolId(): number {
        return {}.PROTOCOL_ID;
    }

    static write(buffer: any, packet: {} | null) {
        if (packet === null) {
            buffer.writeInt(0);
            return;
        }
        {}
    }

    static read(buffer: any): {} | null {
        const length = buffer.readInt();
        if (length === 0) {
            return null;
        }
        const beforeReadIndex = buffer.getReadOffset();
        const packet = new {}();
        {}
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length);
        }
        return packet;
    }
}

export default {};
